src/crypto.rs

```diff
@@ -0,0 +1,206 @@
+    {
+        context_set_property_multiple(self, types::Property::CcmTag, ccm_tag)
+    }
+}
+
+
+/// Initializes the library. Must be called before any other crypto
+/// function. Should be called once in each thread that uses yaca.
+pub fn initialize() -> Result<()>
+{
+    let r = unsafe {
+        lib::yaca_initialize()
+    };
+    conv::res_c_to_rs(r)
+}
+
+/// Cleans up the library.
+/// Must be called before exiting the thread that called yaca_initialize().
+pub fn cleanup()
+{
+    unsafe {
+        lib::yaca_cleanup()
+    }
+}
+
+/// Safely compares first length bytes of two buffers.
+pub fn memcmp(first: &[u8], second: &[u8], length: usize) -> Result<bool>
+{
+    let min = std::cmp::min(first.len(), second.len());
+    assert!(length <= min);
+    let first = first.as_ptr() as *const c_void;
+    let second = second.as_ptr() as *const c_void;
+    let r = unsafe {
+        lib::yaca_memcmp(first, second, length)
+    };
+    conv::res_c_to_rs_bool(r)
+}
+
+/// Generates random data.
+pub fn random_bytes(length: usize) -> Result<Vec<u8>>
+{
+    let mut v: Vec<u8> = Vec::with_capacity(length);
+    let data = v.as_mut_ptr() as *mut c_char;
+    let r = unsafe {
+        lib::yaca_randomize_bytes(data, length)
+    };
+    conv::res_c_to_rs(r)?;
+    unsafe {
+        v.set_len(length);
+    };
+    Ok(v)
+}
+
+fn context_set_property_single<T, U>(ctx: &T, prop: types::Property,
+                                     value: U) -> Result<()>
+    where T: Context + ?Sized,
+{
+    let ctx = ctx.get_handle();
+    let property = conv::property_rs_to_c(&prop);
+    let value: *const U = &value;
+    let value = value as *const c_void;
+    let value_len = mem::size_of::<U>();
+    let r = unsafe {
+        lib::yaca_context_set_property(ctx, property, value, value_len)
+    };
+    conv::res_c_to_rs(r)
+}
+
+// Used by set_property_ccm_aad implementators
+pub(crate) fn context_set_property_multiple<T, U>(ctx: &T, prop: types::Property,
+                                           value: &[U]) -> Result<()>
+    where T: Context + ?Sized,
+{
+    let ctx = ctx.get_handle();
+    let property = conv::property_rs_to_c(&prop);
+    let value_len = value.len() * mem::size_of::<U>();
+    let value = value.as_ptr() as *const c_void;
+    let r = unsafe {
+        lib::yaca_context_set_property(ctx, property, value, value_len)
+    };
+    conv::res_c_to_rs(r)
+}
+
+fn context_get_property_multiple<T, U>(ctx: &T, prop: types::Property) -> Result<Vec<U>>
+    where T: Context + ?Sized,
+          U: Clone,
+{
+    let ctx = ctx.get_handle();
+    let property = conv::property_rs_to_c(&prop);
+    let mut value = ptr::null();
+    let mut value_len = 0;
+    let r = unsafe {
+        lib::yaca_context_get_property(ctx, property, &mut value, &mut value_len)
+    };
+    conv::res_c_to_rs(r)?;
+    let value_len = value_len / mem::size_of::<U>();
+    let v = common::vector_from_raw(value_len, value);
+    Ok(v)
```

Comment:
I might be convinced to change those to 1 liners :-)

Reply:
Would you like me to do some convincing? After all, these are ```{}```-free, just little tiny minute expressions.
<does some convincing>

Reply:
Done